### PR TITLE
working thru serving different content types

### PR DESCRIPTION
**Pull request by julznoob**
Hey Glen,
I worked thru this:
https://www.digitalocean.com/community/tutorials/how-to-create-a-web-server-in-node-js-with-the-http-module

Learned some good stuff. Most of my learnings are in the comments.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,14 +1,52 @@
...
 const hostname = '127.0.0.1';
 const port = 3000;
 
-const server = http.createServer((req, res) => {
-  res.statusCode = 200;
-  res.setHeader('Content-Type', 'text/plain');
-  res.end('Hello World');
+
+// using a read stream. Don't fully understand this
+const htmlPipeFileServer = http.createServer((req, res) => {
```
Comment (julznoob) [inline app.js:7]: I don't fully understand the readstream().pipe() workflow.

Maybe pipe takes a callback? That would make sense. 

Bung all the parsed data into an object and then call `res` with this object.

File: app.js [hunk 2]
```diff
@@ -1,14 +1,52 @@
...
+const jsonServer = http.createServer((req, res) => {
+    res.statusCode = 200;
+    res.setHeader('Content-Type', 'application/json');
+    res.end(`{ "message": "mah messge" }`)
 });
 
-server.listen(port, hostname, () => {
-  console.log(`Server running at http://${hostname}:${port}/`);
+// learnings: 
+// setHeader is used by middleware to add and overwrite headers all willy nilly
```
Comment (julznoob) [inline app.js]: setHeader is used by middleware to add and overwrite headers all willy nilly

But writeHead is like the final, final one and should go just before the res.end is called

File: app.js [hunk 3]
```diff
@@ -1,14 +1,52 @@
...
+
+// But writeHead is like the final, final one and should go just before the res.end is called
+
+const csvServer = http.createServer((req, res) => {
+    res.writeHead(200, {
+        'Content-Type': 'text/csv',
+        'Content-Disposition': 'attachment;filename=goodstuff.csv'
+    })
```
Comment (julznoob) [inline app.js:23]: If i try to call setHeader after writeHead, it will say, "cannot set headers after they are sent to client"

File: app.js [hunk 4]
```diff
@@ -1,14 +1,52 @@
...
+    // will throw: cannot set headers after they are sent to the client
+    // res.statusCode = 200;
+    // res.setHeader('Content-Type', 'text/csv');
+    // res.setHeader('Content-Disposition', 'attachment;filename=goodstuff.csv');
+    res.end(`id,name,email\n1,Happy Plank,plank@forest.com`)
+});
+
+
+// Interesting. The parser will try to close tags if you don't. And random tags dont' break things
+const htmlServer = http.createServer((req, res) => {
```
Comment (julznoob) [inline app.js:27]: Interesting. The parser will try to close tags if you don't. And random tags dont' break things

File: app.js [hunk 5]
```diff
@@ -1,14 +1,52 @@
...
+    res.statusCode = 200;
+    res.setHeader('Content-Type', 'text/html');
+    fs.createReadStream('index.html').pipe(res)
+});
+
+
+// application json. You have to pass the json as a string, which is weird.
+const jsonServer = http.createServer((req, res) => {
+    res.statusCode = 200;
+    res.setHeader('Content-Type', 'application/json');
```
Comment (julznoob) [inline app.js:15]: application json. You have to pass the json as a string, which is weird.

File: app.js (hunk 1 above)
Comment (garrows) [reply app.js:7]: Check out the documentation for pipe. Maybe search for "nodejs streaming tutorial" too.

File: app.js (hunk 5 above)
Comment (garrows) [reply app.js:15]: Yup. It's sent as plain text. But the header above tells the browser to read it as json.

File: app.js (hunk 3 above)
Comment (garrows) [reply app.js:23]: Yeah. Headers need to be sent first so that browser knows how to interpret what's next.

File: app.js (hunk 4 above)
Comment (garrows) [reply app.js:27]: It's not the server that closes the tags. It's the browser trying it's best to figure it that garbage the server sent.

**Review by garrows — APPROVED** [review]
Looking good. Love the experimentation you did. 
Next, try to combine the separate servers you made into one by implementing a "router". Hint: look at req.location (or maybe it's req.path). Can't remember.

**Pull request merged**